core-bundle/src/Security/Voter/DataContainer/PageTypeAccessVoter.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\Voter\DataContainer;
+
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\CoreBundle\Security\DataContainer\CreateAction;
+use Contao\CoreBundle\Security\DataContainer\DeleteAction;
+use Contao\CoreBundle\Security\DataContainer\ReadAction;
+use Contao\CoreBundle\Security\DataContainer\UpdateAction;
+use Doctrine\DBAL\Connection;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AccessDecisionManagerInterface;
+
+/**
+ * @internal
+ */
+class PageTypeAccessVoter extends AbstractDataContainerVoter
+{
+    private const FIRST_LEVEL_TYPES = ['error_401', 'error_403', 'error_404', 'error_503'];
+
+    public function __construct(
+        private readonly AccessDecisionManagerInterface $accessDecisionManager,
+        private readonly Connection $connection,
+    ) {
+    }
+
+    protected function getTable(): string
+    {
+        return 'tl_page';
+    }
+
+    protected function hasAccess(TokenInterface $token, CreateAction|DeleteAction|ReadAction|UpdateAction $action): bool
+    {
+        return $this->validateAccessToPageType($token, $action)
+            && $this->validateFirstLevelType($action)
+            && $this->validateRootType($action);
+    }
+
+    private function validateAccessToPageType(TokenInterface $token, CreateAction|DeleteAction|ReadAction|UpdateAction $action): bool
+    {
+        if ($action instanceof ReadAction) {
+            return true;
+        }
+
+        $types = [];
+
+        if (!$action instanceof CreateAction && isset($action->getCurrent()['type'])) {
+            $types[] = $action->getCurrent()['type'];
+        }
+
+        if (!$action instanceof DeleteAction && isset($action->getNew()['type'])) {
+            $types[] = $action->getNew()['type'];
+        }
+
+        if ([] === $types) {
+            return true;
+        }
+
+        foreach ($types as $type) {
+            if (!$this->accessDecisionManager->decide($token, [ContaoCorePermissions::USER_CAN_ACCESS_PAGE_TYPE], $type)) {
+                return false;
```

Comment:
Again, this breaks the loop at the first type, making it _deny if one type does not match_. It should be _deny if no type matches_ instead, shouldn‘t it?

Reply:
As far as I understand it, for the update action to be allowed, you need access to both, the page type of the current record as well as the page type it would have after updating it.

Reply:
@ausi is correct here as well.

Reply:
I have added a comment in a046814df0a51a8b45de21533f39acb352098fe0.